File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/SubscriptionService.java
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.rest.api.service;

import io.gravitee.common.data.domain.Page;
import io.gravitee.repository.management.api.search.Order;
import io.gravitee.repository.management.model.Subscription;
import io.gravitee.rest.api.model.NewSubscriptionEntity;
import io.gravitee.rest.api.model.SubscriptionEntity;
import io.gravitee.rest.api.model.TransferSubscriptionEntity;
import io.gravitee.rest.api.model.UpdateSubscriptionConfigurationEntity;
import io.gravitee.rest.api.model.UpdateSubscriptionEntity;
import io.gravitee.rest.api.model.common.Pageable;
import io.gravitee.rest.api.model.pagedresult.Metadata;
import io.gravitee.rest.api.model.subscription.SubscriptionMetadataQuery;
import io.gravitee.rest.api.model.subscription.SubscriptionQuery;
import io.gravitee.rest.api.service.common.ExecutionContext;
import java.util.Collection;
import java.util.Map;
import java.util.Set;
import java.util.function.Consumer;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
public interface SubscriptionService {
    /**
     * Get a subscription from its ID.
     * @param subscription
     * @return
     */
    SubscriptionEntity findById(String subscription);

    Collection<SubscriptionEntity> findByApplicationAndPlan(ExecutionContext executionContext, String application, String plan);

    Collection<SubscriptionEntity> findByApi(ExecutionContext executionContext, String api);

    Collection<SubscriptionEntity> findByPlan(ExecutionContext executionContext, String plan);

    SubscriptionEntity create(ExecutionContext executionContext, NewSubscriptionEntity newSubscriptionEntity);

    SubscriptionEntity create(ExecutionContext executionContext, NewSubscriptionEntity newSubscriptionEntity, String customApiKey);

    SubscriptionEntity update(ExecutionContext executionContext, UpdateSubscriptionEntity subscription);

    SubscriptionEntity update(
        ExecutionContext executionContext,
        UpdateSubscriptionEntity updateSubscription,
        Consumer<Subscription> subscriptionTransformer
    );

    SubscriptionEntity update(
        ExecutionContext executionContext,
        UpdateSubscriptionConfigurationEntity updateSubscriptionConfigurationEntity
    );

    SubscriptionEntity updateDaysToExpirationOnLastNotification(String subscriptionId, Integer value);

    SubscriptionEntity pauseConsumer(ExecutionContext executionContext, String subscription);
    SubscriptionEntity pause(ExecutionContext executionContext, String subscription);

    SubscriptionEntity resumeConsumer(ExecutionContext executionContext, String subscriptionId);

    SubscriptionEntity resume(ExecutionContext executionContext, String subscription);

    SubscriptionEntity restore(ExecutionContext executionContext, String subscription);

    SubscriptionEntity fail(String subscriptionId, String failureCause);

    void delete(ExecutionContext executionContext, String subscription);

    Collection<SubscriptionEntity> search(ExecutionContext executionContext, SubscriptionQuery query);

    Page<SubscriptionEntity> search(ExecutionContext executionContext, SubscriptionQuery query, Pageable pageable);

    Page<SubscriptionEntity> search(
        ExecutionContext executionContext,
        SubscriptionQuery query,
        Pageable pageable,
        boolean fillApiKeys,
        boolean fillPlansSecurityType
    );

    Metadata getMetadata(ExecutionContext executionContext, SubscriptionMetadataQuery query);

    SubscriptionEntity transfer(ExecutionContext executionContext, TransferSubscriptionEntity transferSubscription, String userId);

    String exportAsCsv(Collection<SubscriptionEntity> subscriptions, Map<String, Map<String, Object>> metadata);

    Set<String> findReferenceIdsOrderByNumberOfSubscriptions(SubscriptionQuery subscriptionQuery, Order order);

    Set<SubscriptionEntity> findByIdIn(Collection<String> subscriptions);
}

File: gravitee-apim-portal-webui-next/src/services/subscription.service.ts
/*
 * Copyright (C) 2024 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { ConfigService } from './config.service';
import { CreateSubscription, Subscription, SubscriptionStatusEnum, UpdateSubscription } from '../entities/subscription/subscription';
import { SubscriptionsResponse } from '../entities/subscription/subscriptions-response';

@Injectable({
  providedIn: 'root',
})
export class SubscriptionService {
  constructor(
    private readonly http: HttpClient,
    private configService: ConfigService,
  ) {}

  list(queryParams: {
    apiId?: string;
    applicationId?: string;
    statuses: SubscriptionStatusEnum[] | null;
    size?: number;
  }): Observable<SubscriptionsResponse> {
    const params = {
      ...(queryParams.apiId ? { apiId: queryParams.apiId } : {}),
      ...(queryParams.applicationId ? { applicationId: queryParams.applicationId } : {}),
      ...(queryParams.statuses ? { statuses: queryParams.statuses } : { statuses: [] }),
      ...(queryParams.size ? { size: queryParams.size } : {}),
    };
    return this.http.get<SubscriptionsResponse>(`${this.configService.baseURL}/subscriptions`, {
      params,
    });
  }

  get(subscriptionId: string): Observable<Subscription> {
    return this.http.get<Subscription>(
      `${this.configService.baseURL}/subscriptions/${subscriptionId}?include=keys&include=consumerConfiguration`,
    );
  }

  subscribe(createSubscription: CreateSubscription): Observable<Subscription> {
    return this.http.post<Subscription>(`${this.configService.baseURL}/subscriptions`, createSubscription);
  }

  update(subscriptionId: string, updatedSubscription: UpdateSubscription) {
    return this.http.put<Subscription>(`${this.configService.baseURL}/subscriptions/${subscriptionId}`, updatedSubscription);
  }

  resumeConsumerStatus(subscriptionId: string): Observable<Subscription> {
    return this.http.post<Subscription>(`${this.configService.baseURL}/subscriptions/${subscriptionId}/_resumeFailure`, null);
  }
}
